### ISSUE TITLE
YOLOX-tiny support

### ISSUE THREAD

**Issue by ethanlord** [aws/aws-neuron-sdk#403]
Converting YOLOX-tiny model from here: https://github.com/Megvii-BaseDetection/YOLOX
Neuron Compiler version 1.9.1.0+5f09e5c8e
Torch Neuron version 1.10.1.2.1.7.0

When compiling the above model I see the following output

```
INFO:Neuron:All operators are compiled by neuron-cc (this does not guarantee that neuron-cc will successfully compile)
INFO:Neuron:Number of arithmetic operators (pre-compilation) before = 261, fused = 261, percent fused = 100.0%
INFO:Neuron:PyTorch to TF conversion failed to resolve function on aten::meshgrid with inputs [[<tf.Tensor 'range:0' shape=(52,) dtype=int32>, <tf.Tensor 'range_1:0' shape=(52,) dtype=int32>], 'ij']
INFO:Neuron:Exception = meshgrid() takes 2 positional arguments but 3 were given
```

This is fixable by removing `indexing="ij"` from all meshgrid operators, but leads the following warning from pytorch which likely indicates a problem in the future:

```
UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2157.)
```

After that fix, I see the following error:

```
INFO:Neuron:All operators are compiled by neuron-cc (this does not guarantee that neuron-cc will successfully compile)
INFO:Neuron:Number of arithmetic operators (pre-compilation) before = 261, fused = 261, percent fused = 100.0%
INFO:Neuron:Compiling function _NeuronGraph$847 with neuron-cc
INFO:Neuron:Compiling with command line: '/home/ubuntu/.local/bin/neuron-cc compile /tmp/tmp09o6_spy/graph_def.pb --framework TENSORFLOW --pipeline compile SaveTemps --output /tmp/tmp09o6_spy/graph_def.neff --io-config {"inputs": {"tensor.1:0": [[1, 3, 416, 416], "float32"]}, "outputs": ["transpose_259:0"]} --verbose 35'
.03/08/2022 11:54:52 PM ERROR 23886 [neuron-cc]: Failed to parse model /tmp/tmp09o6_spy/graph_def.pb: The following operators are not implemented: {'SelectV2'} (NotImplementedError)
```

Any advice on how to proceed from here?

**Comment by ethanlord** [comment]
This was fixed by rerunning `pip install neuron-cc[tensorflow]`

**Comment by ethanlord** [comment]
Accidentally closed this prematurely, the model successfully converts but produces dramatically different outputs. Using the repo mentioned above and the yolox-tiny weights found in the readme, this can be reproduced with the following code:

```
from yolox.exp import get_exp
from yolox.utils import fuse_model
from yolox.data.data_augment import ValTransform
import torch
import torch_neuron
import cv2
import numpy as np

exp = get_exp(None, 'yolox-tiny')
model = exp.get_model()

model.eval()
ckpt = torch.load('./yolox_tiny.pth', map_location="cpu")
# load the model state dict
model.load_state_dict(ckpt["model"])
model = fuse_model(model)

img = torch.zeros(1, 3, *exp.test_size)

res = model(img)

model_neuron = torch.neuron.trace(model, img)

res_neuron = model_neuron(img)
```

**Comment by aws-taylor** [comment]
Hello @ethanlord,

 I'm working to reproduce the issue. Could I request that you run` python -m pip freeze` and include the results so that I am better able to mimic your test environment.  Compilation results can change depending on third-party packages such as torchvision, h5py, pytorch, transformers, ...

Regards,
Taylor

**Comment by ethanlord** [comment]
Sure, here is the output from that

```
absl-py==0.15.0
aiocontextvars==0.2.2
alabaster==0.7.12
antlr4-python3-runtime==4.8
appdirs==1.4.3
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
asn1crypto==0.24.0
astor==0.8.1
astroid==2.9.3
astunparse==1.6.3
async-generator==1.10
attrs==21.4.0
Automat==20.2.0
awscli==1.22.43
Babel==2.9.1
backcall==0.2.0
bleach==4.1.0
blinker==1.4
bokeh==2.3.3
boto==2.49.0
boto3==1.20.43
botocore==1.23.43
cached-property==1.5.2
cachetools==4.2.4
certifi==2021.10.8
cffi==1.15.0
chardet==3.0.4
charset-normalizer==2.0.10
clang==5.0
click==8.0.3
cloud-init==21.4
cmake==3.22.1
colorama==0.4.3
command-not-found==0.3
configobj==5.0.6
constantly==15.1.0
contextvars==2.4
coremltools==5.1.0
cpplint==1.5.5
cryptography==36.0.1
cycler==0.11.0
dataclasses==0.8
decorator==4.4.2
defusedxml==0.7.1
distro-info===0.18ubuntu0.18.04.1
dmlc-nnvm==1.9.1.0+0
dmlc-topi==1.9.1.0+0
dmlc-tvm==1.9.1.0+0
docutils==0.15.2
dparse==0.5.1
ec2-hibinit-agent==1.0.0
entrypoints==0.3
environment-kernels==1.1.1
flatbuffers==1.12
future==0.18.2
gast==0.2.2
google-auth==1.35.0
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
graphviz==0.19.1
grpcio==1.44.0
h5py==3.1.0
hibagent==1.0.1
httplib2==0.20.2
hyperlink==21.0.0
idna==3.3
imageio==2.14.1
imagesize==1.3.0
immutables==0.16
importlib-metadata==4.8.3
incremental==21.3.0
inferentia-hwm==1.9.0.0+0
ipykernel==5.5.6
ipython==7.16.3
ipython-genutils==0.2.0
ipywidgets==7.6.5
islpy==2018.2+aws2018.x.891.0.bld0
isort==5.10.1
jedi==0.17.2
Jinja2==3.0.3
jmespath==0.10.0
joblib==1.1.0
jsonpatch==1.16
jsonpointer==1.10
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==7.1.2
jupyter-console==6.4.0
jupyter-core==4.9.1
jupyterlab-pygments==0.1.2
jupyterlab-widgets==1.0.2
keras==2.6.0
Keras-Applications==1.0.8
keras-nightly==2.5.0.dev2021032900
Keras-Preprocessing==1.1.2
kiwisolver==1.3.1
language-selector==0.1
lazy-object-proxy==1.7.1
loguru==0.6.0
Markdown==3.3.6
MarkupSafe==2.0.1
matplotlib==3.3.4
mccabe==0.6.1
mistune==0.8.4
mpmath==1.2.1
nbclient==0.5.9
nbconvert==6.0.7
nbformat==5.1.3
nest-asyncio==1.5.4
netifaces==0.10.4
networkx==2.4
neuron-cc==1.9.1.0+5f09e5c8e
ninja==1.10.2.3
nose==1.3.7
notebook==6.4.8
numpy==1.19.5
nvidia-ml-py==11.450.51
oauthlib==3.2.0
omegaconf==2.1.1
onnx==1.8.1
onnx-simplifier==0.3.5
onnx-tf==1.9.0
onnx2pytorch==0.4.1
onnxoptimizer==0.2.6
onnxruntime==1.10.0
opencv-python==4.5.5.62
opt-einsum==3.3.0
packaging==21.3
PAM==0.4.2
pandas==1.1.5
pandocfilters==1.5.0
parso==0.7.1
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.4.0
platformdirs==2.4.0
prometheus-client==0.13.0
prompt-toolkit==3.0.24
protobuf==3.19.3
ptyprocess==0.7.0
pyasn1==0.4.8
pyasn1-modules==0.2.1
pycocotools==2.0.4
pycparser==2.21
pycrypto==2.6.1
pydot==1.4.2
pygal==3.0.0
Pygments==2.11.2
PyGObject==3.26.1
PyJWT==1.5.3
pylint==2.12.2
pyOpenSSL==17.5.0
pyparsing==3.0.7
pyrsistent==0.18.0
pyserial==3.4
pystache==0.6.0
python-apt==1.6.5+ubuntu0.7
python-dateutil==2.8.2
python-debian==0.1.32
pytz==2021.3
PyWavelets==1.1.1
PyYAML==5.4.1
pyzmq==22.3.0
qtconsole==5.2.2
QtPy==2.0.0
requests==2.27.1
requests-oauthlib==1.3.1
requests-unixsocket==0.1.5
rsa==4.7.2
s3transfer==0.5.0
safety==1.10.3
scikit-image==0.17.2
scikit-learn==0.24.2
scipy==1.4.1
seaborn==0.11.2
Send2Trash==1.8.0
service-identity==16.0.0
six==1.15.0
snowballstemmer==2.2.0
sos==4.1
Sphinx==4.4.0
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
ssh-import-id==5.7
sympy==1.9
systemd-python==234
tabulate==0.8.9
tensorboard==1.15.0
tensorboard-data-server==0.6.1
tensorboard-plugin-neuron==2.2.0.0
tensorboard-plugin-wit==1.8.1
tensorflow==1.15.0
tensorflow-estimator==1.15.1
termcolor==1.1.0
terminado==0.12.1
testpath==0.5.0
thop==0.0.31.post2005241907
threadpoolctl==3.0.0
tifffile==2020.9.3
timm==0.5.4
toml==0.10.2
torch==1.10.1
torch-neuron==1.10.1.2.1.7.0
torchvision==0.11.2
tornado==6.1
tqdm==4.62.3
traitlets==4.3.3
Twisted==21.7.0
typed-ast==1.5.2
typeguard==2.13.3
typing-extensions==3.7.4.3
ubuntu-advantage-tools==27.5
ufw==0.36
unattended-upgrades==0.1
urllib3==1.26.8
wcwidth==0.2.5
webencodings==0.5.1
Werkzeug==2.0.3
widgetsnbextension==3.5.2
wrapt==1.12.1
-e git+ssh://git@github.com/Megvii-BaseDetection/YOLOX.git@5bbfc110d73e80df4406e96b3c1103bb3555770f#egg=yolox
zipp==3.6.0
zope.interface==5.4.0
```

**Comment by aws-taylor** [comment]
Hello again @ethanlord,

 We were able to reproduce the issue and found that by using `--fp32-cast=matmult-fp16 --fast-math=no-fast-relayout` we were able to reduce the numeric mismatch quite a bit.  The documentation on https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-cc/command-line-reference.html for `--fast-math=none` is incorrect (we're fixing it). 
 
 Regards,
Taylor

**Comment by aws-diamant** [comment]
Hi @ethanlord ,

Just to give you a quick update - we're actively working on this, and have narrowed down the problem to lowering of a specific operator.
We'll update you as soon as we have further details.

**Comment by hannanjgaws** [comment]
Hi @ethanlord,

We have updated our `meshgrid` implementation to support the new `indexing` argument. This change will come out in an upcoming release.

You can fix the accuracy issue you’re seeing by removing the in-place slice assignments in the `decode_outputs` function. The fix is the following:
```
diff --git a/yolox/models/yolo_head.py b/yolox/models/yolo_head.py
index d3ca5e6..72a3d1a 100644
--- a/yolox/models/yolo_head.py
+++ b/yolox/models/yolo_head.py
@@ -246,8 +246,9 @@ class YOLOXHead(nn.Module):
         grids = torch.cat(grids, dim=1).type(dtype)
         strides = torch.cat(strides, dim=1).type(dtype)
 
-        outputs[..., :2] = (outputs[..., :2] + grids) * strides
-        outputs[..., 2:4] = torch.exp(outputs[..., 2:4]) * strides
+        output_1 = (outputs[..., :2] + grids) * strides
+        output_2 = torch.exp(outputs[..., 2:4]) * strides
+        outputs = torch.cat((output_1, output_2, outputs[..., 4:]), -1)
         return outputs
 
     def get_losses(
```
We are investigating ways to automate the process of replacing in-place slicing assignments, but for now we recommend you make the change locally in the model source code.

**Comment by ethanlord** [comment]
Thanks, the outputs all look correct now. Is there a reason why this model would run slower than something like resnet50? We're seeing average throughput of around 500 images/second with `torch.neuron.DataParallel` after placing slice operations on the CPU like in the [YoloV4 tutorial](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/yolo_v4.html#Run-compilation-with-manually-specified-device-placement), despite this model being much smaller than resnet50.

**Comment by hannanjgaws** [comment]
Thanks for confirming that the outputs are correct! 

In terms of performance: The arithmetic intensity varies from model to model, which can explain performance differences between different models. What other YOLOX-tiny throughput numbers are you comparing to? Also, are you able to provide the script you're using to benchmark your model on inf1?

**Comment by ethanlord** [comment]
We ran some more benchmarks across different batch sizes and also with yolox-nano. We just saw that CPU throughput was similar or faster than CPU performance for resnet50, but when running on inf1 these models have lower throughput than the figures you report for resnet50 [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/benchmark/index.html). Here is the benchmark script we used:

```
from yolox.exp import get_exp
from yolox.utils import fuse_model
import torch
import torch_neuron
import numpy as np
import time

for model_size in ['tiny', 'nano']:
    exp = get_exp(None, f'yolox-{model_size}')
    model = exp.get_model()

    img = torch.zeros(1, 3, *exp.test_size).float()

    model.eval()
    ckpt = torch.load(f'./yolox_{model_size}.pth', map_location="cpu")
    # load the model state dict
    model.load_state_dict(ckpt["model"])
    model = fuse_model(model)

    for batch_size in [1, 4, 8, 12, 16]:
        print(f'Model: yolox-{model_size}, Batch Size: {batch_size}')
        batch_image = torch.tile(img, (batch_size, 1, 1, 1))

        start = time.time()
        num_iter = 10
        for i in range(num_iter):
            model(batch_image)
        end = time.time()
        print(f'CPU Throughput: {num_iter*batch_size/(end-start)}/sec')

        def subgraph_builder_function(node):
            return 'slice' not in node.name

        model_neuron = torch.neuron.trace(model, batch_image, subgraph_builder_function=subgraph_builder_function, strict=False)
        torch.jit.save(model_neuron, f'yolox_{model_size}_neuron_b{batch_size}.pt')
        model_neuron = torch.neuron.DataParallel(torch.jit.load(f'yolox_{model_size}_neuron_b{batch_size}.pt'))

        # Preload model
        model_neuron(batch_image)

        test_size = 40*batch_size
        batch_image = torch.tile(img, (test_size, 1, 1, 1))

        start = time.time()
        num_iter = 20
        for i in range(num_iter):
            model_neuron(batch_image)
        end = time.time()
        print(f'Neuron Throughput: {test_size*num_iter/(end-start)}/sec')
```

**Comment by ethanlord** [comment]
Update on this, we were able to achieve the throughput we wanted by decreasing input resolution, I don't think we have any more concerns here.

**Comment by hannanjgaws** [comment]
Great to hear you were able to meet your performance goals on Neuron. I will close this ticket, but please feel free to reopen it if you have any additional questions about running this model on inf1.
